### ISSUE TITLE
A number of run mode issues

### ISSUE THREAD

**Issue by CxRes** [nwjs-community/nw-builder#482]
I found a number of somewhat related issues with run mode and am opening them together, since its easier for me:

1) When a default platform is not specified from the cli in **run** mode, NW-builder fails trying to download `osx32` binaries. When I dug into the code I found that by default is set as `['osx32','osx64', 'win32', 'win64']` in cli and ['osx64', 'win32', 'win64'] in the JS api. Now I think in general the default platform should be your current platform, given that you have a decently reliable way of detecting it. But this is a problem in **run** mode as it will try to download binaries which one has absolutely no need for.

2) An `argv` argument on the `nw` object is set presumably to send arguments to NW.js instance at runtime. However since the arguments are being retrieved using `process.argv.slice(2)`, this fails when using the cli version as the nwbuild flags are also passed to NW.js. A better strategy imho is to accept arguments after a `--` flag much like `npm run` does.

3) The `spawn` for NW.js contains `'--enable-logging'` option hardcoded which produces ugly terminal windows (at least on Windows). If a user needs this they should pass it by hand - such as by using `--` option.

Before I make a PR, I would like someone to check if the fixes I am suggesting are OK. Please check them out at https://github.com/CxRes/nw-builder/commit/b42d5ef14fc303843755b70ab7117346fdd5d12f

**Comment by zaygraveyard** [comment]
@CxRes thank you for your work
I don't think asking "Anybody Home???!!!???" will get them to find time to reply to you any faster
I'll post some ideas that are in the same spirit, might help the discussion

**Comment by zaygraveyard** [comment]
* I would drop `'osx32'` since the last NW.js version to support it is `v0.12.3`.
* I would not change the default `-platforms` to the `<Current OS>`, as this change might be considered backwards incompatible.
* As for the `--` option, it should be already supported see [optimist's stop parsing](https://github.com/substack/node-optimist#stop-parsing)

**Comment by CxRes** [comment]
@zaygraveyard  Thanks for the encouragement :-P. You are the only one who has responded :) !!!!!!!

The trouble is that there have been no updates for last few months. At least mentioning the maintainers by name puts it in their notification (There is a failure actually on part of the actual NW.js crew to realize the importance of critical tooling - they must not just leave it to the community. Despite a faster update cycle, NW.js has lost traction to Electron as a result).

While I am not married to the change, how likely is it that changing `--platforms` to <current OS> will break builds. Only thing that will happen is a few missing build that do not run on the host OS anyway. Besides, that is what a major version bump is for. All that said, one does not even need to to touch `--platforms` when invoking run, I just temporarily save it to another variable. Only drawback is one cannot run `.build()` and `.run()` in parallel.

`--` parsed value is not passed to the `argv` variable in nw-builder options. I am aware of optimist stop parsing but that will not work because the `argv._` will also include file paths (ie undashed arguments prior to `--`) that we already pass as the `parentDirectory` when spawning NW.js.

**Comment by zaygraveyard** [comment]
According to the [contributors graph](https://github.com/nwjs-community/nw-builder/graphs/contributors) @adam-lynch is the only active maintainer of this repo.
And @rogerwang with @GnorTech are the only active maintainers of NW.js (see [contributors graph of nw.js](https://github.com/nwjs/nw.js/graphs/contributors)).
So when talk about the NW.js crew it's mainly these two guys, as far as I can tell.
I say this to show you that there are very few people behind these projects and they don't always have free time on their hands.
And that's all the more reason for us to contribute, but we should give them a bit of time to reply and review the code.

And to honest, the only reason I'm still using NW.js instead of Electron is that the project that uses it (https://github.com/oklai/koala) was started before Electron was a thing and now I don't have enough free time to migrate it.
Plus, a big reason for me to replace NW.js with Electron is, as you said, tooling :disappointed:.

With that out the way, lets get back to your changes.

It's up to the maintainers to decide if the `--platforms` change justifies a major version bump.
The `--run` uses the current os by default (if no `--platfroms` is specified)
And without the `--run`, it will just build so it doesn't need to run them.

As for `--`, you are right but I think we can pass `argv._.slice(1)` as the `argv` option to `NWBuilder` (since only one undashed argument is acceptable and required, ie. the file)

**Comment by CxRes** [comment]
@zaygraveyard I agree with you absolutely. It is a chicken and the egg problem. People are leaving NW.js because there is no tooling and then there are not enough people left to contribute. Any big enough project will have to make judicious use of resources and it is a very tough act. But, I do think NW.js somewhere did get their priorities wrong. Having said that I wish them well as we need the competition of ideas and @rogerwang will always be the pioneer of this space.

To the changes:
1. Agreed on the bump, that is why I asked for a review. But `--run` will download binaries based on the values in `options.platform`  or default `--platforms` (resp. on API and CLI versions) i.e. it starts downloading OS X binaries on Windows and vice-versa even when using the `--run` flag. This behaviour is wrong!
2. We can do what you are saying as well! I just prefer my thing as it is a little more tolerant. It wont pass extra args if the user accidentally writes a second or third arg before the dash. And, its just one extra line of code.

**Comment by adam-lynch** [comment]
@zaygraveyard thanks for stepping in here. Yeah, finding time to work on stuff like this is tough. I'm always looking to on-board new contributors / maintainers. And yes, Roger Wang is king 👑.

As far as the `osx32` issue goes, I think #469 was intended to fix that but we could close it if there's a new PR. 

So I'm reading what has been said correctly, `.run` correctly defaults to the current platform but the problem is it downloads unnecessary binaries for other platforms. Yeah that sounds like a bug alright, but no need for a major bump, unless I'm misunderstanding something.

About `--enable-logging`, we do have a `--quiet` option to turn off logging. Maybe that should determine whether or not `--enable-logging` is passed?

**Comment by CxRes** [comment]
@adam-lynch Good to hear from you... Sorry if I am being a little "persistent"!

re: `--platforms`: The bug is limited to download of binaries for other platforms. But imho having a fixed choice for what OS you want to build for is wrong. To my mind it makes more sense that default should be to build specific to your own OS. I would imagine someone on Linux certainly does not want to trigger OSX and Windows builds. 

Also the default options for `--platforms` in the CLI and API are different (API version still has `osx32`) - which is definitely a bug. Unless there is a specific override expected for the CLI, it would be more DRY to define defaults just once (i.e. in the API only)

re `--enable-logging`: It is the only option we pass when invoking the NW executable which I felt was bit arbitrary. Though, if you desire this behaviour to be predicated on the `--quiet` flag (I would personally prefer `--verbose` flag but then I do not want to break the API), that is something I can easily fix.

**Comment by zaygraveyard** [comment]
@adam-lynch I'm happy to help anyway I can :smile: 

#469 also adds a `--name` option to the CLI so this change doesn't replace it entirely.
re `--enable-logging`: I have the same preference as @CxRes

**Comment by kessler** [comment]
@CxRes I was under the impression that https://github.com/nwjs-community/nw-builder/pull/469 is the PR related to this issue... If I was mistaken then it's no biggie we'll just reopen the issue.

**Comment by CxRes** [comment]
@kessler #469 is related but does not address the things I have raised (see discussion above). I would strongly encourage you to review CxRes/nw-builder@b42d5ef so that I can put up a PR.

**Comment by kessler** [comment]
note to self: waiting for response on related PR

**Comment by ayushmxn** [comment]
Fixed by #488
